agent/api/task/service_connect_validation.go

```diff
@@ -0,0 +1,293 @@
+			numOfFoundSCContainer += 1
+		}
+	}
+
+	if numOfFoundSCContainer == 0 {
+		return fmt.Errorf("service connect container name=%s does not exist in the task", scContainerName)
+	} else if numOfFoundSCContainer > 1 {
+		return fmt.Errorf("found %d duplicate service connect container name=%s exists in the task", numOfFoundSCContainer, scContainerName)
+	}
+
+	return nil
+}
+
+// validateEgressConfig validates the service connect egress config.
+func validateEgressConfig(scEgressConfig *EgressConfig, ipv6Enabled bool) error {
+	// egress config can be empty for the first service since there are no other tasks that it can talk to
+	if scEgressConfig == nil {
+		return nil
+	}
+
+	// ListenerName is a required field if the egress config is not empty
+	if scEgressConfig.ListenerName == "" {
+		return fmt.Errorf("no service connect listener name")
+	}
+
+	// VIP is a required field if the egress config is not empty
+	if !ipv6Enabled && scEgressConfig.VIP.IPV4CIDR == "" {
+		return fmt.Errorf("no service connect VIP IPV4CIDR")
+	}
+
+	if ipv6Enabled && scEgressConfig.VIP.IPV6CIDR == "" {
+		return fmt.Errorf("no service connect VIP IPV6CIDR. IPV6CIDR must not be empty when the task is IPV6 enabled")
+	}
+
+	// validate IPV4CIDR if it exists
+	if scEgressConfig.VIP.IPV4CIDR != "" {
+		trimmedIpv4cidr := strings.TrimSpace(scEgressConfig.VIP.IPV4CIDR)
+		if err := validateCIDR(trimmedIpv4cidr, "IPv4 CIDR"); err != nil {
+			return err
+		}
+	}
+
+	// validate IPV6CIDR if it exists
+	if scEgressConfig.VIP.IPV6CIDR != "" {
+		trimmedIpv6cidr := strings.TrimSpace(scEgressConfig.VIP.IPV6CIDR)
+		if err := validateCIDR(trimmedIpv6cidr, "IPv6 CIDR"); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// validateDnsConfig validates the service connnect DNS config.
+func validateDnsConfig(scDnsConfligList []DNSConfigEntry, scEgressConfig *EgressConfig) error {
+	if len(scDnsConfligList) == 0 && scEgressConfig != nil {
+		// DNS config is a field which associates to egress config
+		return fmt.Errorf("no service connect DNS config. The DNS config is required when the egress config is not empty")
+	}
+
+	for _, dnsEntry := range scDnsConfligList {
+		// HostName is a required field
+		if dnsEntry.HostName == "" {
+			return fmt.Errorf("no hostname in the DNS config entry hostname=%s, address=%s", dnsEntry.HostName, dnsEntry.Address)
+		}
+
+		// Address is a required field
+		if dnsEntry.Address == "" {
+			return fmt.Errorf("no address in the DNS config entry hostname=%s, address=%s", dnsEntry.HostName, dnsEntry.Address)
+		}
+
+		// validate the address is a valid IPv4/IPv6 address
+		if err := validateAddress(dnsEntry.Address); err != nil {
+			return fmt.Errorf("invalid address in the DNS config entry hostname=%s, address=%s: %w", dnsEntry.HostName, dnsEntry.Address, err)
+		}
+	}
+
+	return nil
+}
+
+// validateAddress validates the passed address is a valid IPv4/IPv6.
+func validateAddress(address string) error {
```

Comment:
Would recommend checking against the given protocol so that we don't take an IPv4 inside a supposedly IPv6 field.
I think we can do something similar to Rajal's [PR](https://github.com/aws/amazon-vpc-cni-plugins/pull/64/files#diff-6ecce0879e112f7e1e1d78bdf2d8ac8639491588157e561563df634554a12dbc) (i.e. using To4() and To16())

Reply:
Discussed offline, as DNS config comes in as a list, we will validate CIDR based on the given protocol.